### ISSUE TITLE
Supporting filesystem mount options

### ISSUE THREAD

**Issue by rootfs** [kubernetes/kubernetes#31613]
This is a generic discussion on how should we support filesystem mount options. 

We currently support network filesystems: NFS, Glusterfs, Ceph FS, SMB (Azure file), Quobytes. 
Mount time options that are operationally important and have no security implications should be suppported. Examples are NFS's TCP mode, versions, lock mode, caching mode; Glusterfs's caching mode; SMB's version, locking, id mapping; and more.

Mount time options for local filesystems such as ext4 and xfs are not discus, yet, but should benefit from the solutions to this issue.

One solution is to walk through all the mount options and create API objects for each one of them in their volume source. This solution clearly states what are the supported options. But supporting all valid mount options makes us liable to compatibility issues when mount options changes or between OSes (Linux, BSD/OSX)

Second solution also adds an API objects in volume source. Instead of 1-1 mapping, the single API object is a string comma separated blob like `version=3,protoc=tcp,nolock`. It is the convention used by mount command. Since this solution only supported a subset of filesystem options, it is not subject to mount option changes. The counter argument is that end users are not immediately aware of what options are supported. 

Based on discussions in #24721 and #17226(tcp and lock). 

@kubernetes/sig-storage

**Comment by rootfs** [comment]
@saad-ali @childsb can we spend time on it during sig meeting?

**Comment by sferich888** [comment]
Does https://github.com/kubernetes/kubernetes/issues/17226 need to be linked to this too?

**Comment by sayash** [comment]
Not sure if directly relevant, but adding mount options to volumes may also help in setting tmpfs / shm sizes also. I'm currently using the method outlined in [openshift documentation](https://docs.openshift.org/latest/dev_guide/shared_memory.html) for mounting a larger SHM into a pod's container than the default 64 MB given by Docker. Being able to set a mount options such as `size` could help here. Maybe this is another way to do this : pull request #34928.

**Comment by thockin** [comment]
How many FS'es besides NFS have commonly used mount options?

In general, Linux is pretty good about not breaking those sorts of things, I think...  Once we add it, we can count on it being there...

**Comment by rootfs** [comment]
@thockin there are a few :)
-  CIFS. There are important options such as smb version. Currently in azure_file plugin, [mount options](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_file/azure_file.go#L198) are fixed. But I've seen requests to use different smb version (2.1 rather than 3.0). 
- FUSE. There are fuse options (allow_other, etc) that are frequently overridden. 
- Gluster. We are planning providing a new option to present mount time failures to console other than dumping errors in logs. This needs a mount time option.

**Comment by kobel169** [comment]
For example I need to "discard" option to mount the rbd storage with ext4. 
Without it, I can not appreciate the scale of cluster ceph really busy volume, and I can not be sure of the results ceph du.

**Comment by thockin** [comment]
So this is a tradeoff. Allowing arbitrary strings is flexible but forces users to be aware of mechanism.  Crystalizing into API is better UX, but means we need to be responsive.

opinions?  @saad-ali @matchstick @jingxu97 @kubernetes/sig-storage

**Comment by eparis** [comment]
If we are going to support flex/unknown volume/fs types do we have any option other than 'free form' ?

Also, these things "should" hopefully only be set by a storage admin who actually knows about the underlying storage...

**Comment by thockin** [comment]
Well, moving to flex/external drivers makes static API validation harder,
it's not impossible - you could call the driver's "validate" call from
apiserver, for example.  But even if we don't do that, there's a difference
between:

```
    volumes:
      - name: foo
        nfs:
            softErrors: true
            interruptible: true
            readSize: 42
            server: 1.2.3.4
```

and

```
    volumes:
      - name: foo
        mountFlags:
          - soft
          - intr
          - rszise=42
        nfs:
            server: 1.2.3.4
```

**Comment by eparis** [comment]
Closing this issue as this is resolved in 1.6. See:
https://github.com/kubernetes/kubernetes/pull/41906
https://github.com/kubernetes/kubernetes.github.io/pull/2743

**Comment by wied03** [comment]
@sayash - I don't think the PR above addresses the tmpfs issue fully. The `sizeLimit` field in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#emptydirvolumesource-v1-core might do what you were talking about. There are no general mount options for volumes though from what I can tell and the `mountOptions` support is only for persistent volumes (and emptyDir isn't on the list).
